msg-socket/src/pub/mod.rs

```diff
@@ -0,0 +1,197 @@
+use bytes::Bytes;
+use msg_wire::pubsub;
+use thiserror::Error;
+
+mod driver;
+mod session;
+mod socket;
+mod stats;
+mod trie;
+pub use socket::*;
+use stats::SocketStats;
+
+#[derive(Debug, Error)]
+pub enum PubError {
+    #[error("IO error: {0:?}")]
+    Io(#[from] std::io::Error),
+    #[error("Wire protocol error: {0:?}")]
+    Wire(#[from] msg_wire::reqrep::Error),
+    #[error("Authentication error: {0}")]
+    Auth(String),
+    #[error("Socket closed")]
+    SocketClosed,
+    #[error("Topic already exists")]
+    TopicExists,
+    #[error("Unknown topic: {0}")]
+    UnknownTopic(String),
+    #[error("Topic closed")]
+    TopicClosed,
+    #[error("Transport error: {0:?}")]
+    Transport(#[from] Box<dyn std::error::Error + Send + Sync>),
+}
+
+#[derive(Debug)]
+pub struct PubOptions {
+    pub max_connections: Option<usize>,
+    pub session_buffer_size: usize,
+    /// The interval at which each session should be flushed. If this is `None`,
+    /// the session will be flushed on every publish, which can add a lot of overhead.
+    pub flush_interval: Option<std::time::Duration>,
+}
```

Comment:
what do you think of supporting pubsub synchronization? 
Could easily be a separate issue as it's an extra feature.

Basically let the pub start sending only when all `expected_subscribers` are connected and subscribed.
This is useful if you want to make sure that subscribers don't lose data if they are slow to startup.

- https://zguide.zeromq.org/docs/chapter2/#Node-Coordination

Reply:
I like this, but maybe this can also be solved by having "durable" subscribers like in NATS. Would have to change the buffering methods a little bit (as right now, the only option is to start receiving messages from the moment a session is spawned due to the mechanics of the broadcast channel), but the idea is that all buffered messages will be sent to any new subscribers, and the sessions keep track of the sequence numbers (through an ack mechanism). This way, we guarantee no data loss (unless the subscriber is so slow to connect that some buffered items are already evicted).

Reply:
In my opinion this behaviour should be configurable, as in some cases this is not expected to happen.
Example: if I subscribe to Fiber's `pending_transactions` I only want to listen to future messages ideally.

This durability could also be a challenge for high-throughput publishers, as the request queue would be kept in memory which is also limited

Reply:
100% agree. Wether or not the subscriber is durable will be decided on the subscriber's end by setting a subscription policy. However, the limits should be set on the publisher side. We can start to do it per session, but ideally we'd have it per topic in the future. Can also imagine custom storage backends for topics (i.e. rocksdb disk instead of in-memory LRU)

Reply:
- Added this and some other details in https://github.com/chainbound/msg-rs/issues/13